### PR TITLE
Update 1 Introduction notebooks for v24

### LINKED ISSUE #155: Update to CIL version 24.2 (latest release)
Also check 
- https://github.com/TomographicImaging/CIL/pull/1826
- Update demos for changes with CIL >= [v24.2.0](https://github.com/TomographicImaging/CIL/blob/v24.2.0/CHANGELOG.md)
- All notebooks with iterative algorithms, check different API, e.g. use of `max_iterations` and `run`
- `CGLS` removed `tolerance`
- `GC` and `Armijo` rule

Notebooks working correctly with latest [release](https://github.com/TomographicImaging/CIL/releases):
- [x] 1_Introduction @hrobarts #162 
  + [x] 05_usb_limited_angle_fbp_sirt.ipynb
- [x] 2 @MargaretDuff
- [x] 3_Multichannel @casperdcl #160

### PR DESCRIPTION
**Pull request by hrobarts**
Check the CIL-Demos notebooks in the Introduction folder work with CIL v24.1

- [x] 00_CIL_geometry.ipynb
- [x] 01_intro_walnut_conebeam.ipynb
- [x] 02_intro_sandstone_parallel_roi.ipynb
- [x] 03_preprocessing.ipynb
- [x] 04_FBP_CGLS_SIRT.ipynb 
    - Remove max iterations and tolerance, change example to checking convergence with a plot of the objective value
- [x] 05_usb_limited_angle_fbp_sirt.ipynb
     - Remove max iterations

### CODE REVIEW

**Comment by MargaretDuff** [comment]
@casperdcl  - any idea why nothing is showing when CGLS is run in notebook folder 1 number 4? 
image
I think it is important to fix this because we ask the user to try running the cell twice to see what happens but currently nothing is happening!

File: demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb [hunk 1]
```diff
@@ -402,7 +398,14 @@
    "outputs": [],
    "source": [
     "# run N interations\n",
```
Comment (lauramurgatroyd) [inline demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb]:
```suggestion
    "# run N iterations\n",
```

File: demos/1_Introduction/05_usb_limited_angle_fbp_sirt.ipynb [hunk 2]
```diff
@@ -101,7 +101,7 @@
    "outputs": [],
    "source": [
     "# Please set the filename yourself, if you are running the notebook locally:\n",
-    "filename = \"/mnt/materials/SIRF/Fully3D/CIL/Usb/gruppe 4_2014-03-20_1404_12/tomo-A/gruppe 4_tomo-A.txrm\"\n",
+    "filename = \"/mnt/share/materials/SIRF/Fully3D/CIL/Usb/gruppe 4_2014-03-20_1404_12/tomo-A/gruppe 4_tomo-A.txrm\"\n",
```
Comment (lauramurgatroyd) [inline demos/1_Introduction/05_usb_limited_angle_fbp_sirt.ipynb]:
```suggestion
    "filename = \"/mnt/materials/SIRF/Fully3D/CIL/Usb/gruppe 4_2014-03-20_1404_12/tomo-A/gruppe 4_tomo-A.txrm\"\n",
```

Please revert the change as this matches the path to data on jupyterhub

File: demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb [hunk 3]
```diff
@@ -659,6 +678,13 @@
     "       ['FBP, noisy projections, difference from phantom', 'CGLS, noisy projection, difference from phantom', 'SIRT, noisy projection, difference from phantom'], \\\n",
     "       cmap=cmap, num_cols=2, size=(15,15), origin='upper-left', fix_range=True)"
    ]
+  },
+  {
+   "cell_type": "code",
```
Comment (lauramurgatroyd) [inline demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb]: Looks like there is an empty cell at end of notebook? Please remove

File: demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb
Comment (lauramurgatroyd) [inline demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb:1]: Please add your name to authors

File: demos/1_Introduction/05_usb_limited_angle_fbp_sirt.ipynb
Comment (lauramurgatroyd) [inline demos/1_Introduction/05_usb_limited_angle_fbp_sirt.ipynb:1]: Please add your name to authors

File: demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb [hunk 4]
```diff
@@ -368,7 +367,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Alternatively, tolerance can be used as a stopping criterion."
+    "If we run a lot of iterations we can check if the algorithm converges. We set the `tolerance` to 0 to demonstrate the algorithm continuing to run (but this way of setting tolerance is being deprecated)."
```
Comment (lauramurgatroyd) [inline demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb:372]: What will we use instead in the future in terms of the tolerance? Sorry I've forgotten

**Review by lauramurgatroyd — CHANGES_REQUESTED** [review]
Looking good, just some small changes requested

**Comment by lauramurgatroyd** [comment]
> @casperdcl - any idea why nothing is showing when CGLS is run in notebook folder 1 number 4? image I think it is important to fix this because we ask the user to try running the cell twice to see what happens but currently nothing is happening!

Just a note - this worked fine for me - I was able to see the progress bar both times I ran the cell

**Comment by casperdcl** [comment]
> Just a note - this worked fine for me - I was able to see the progress bar both times I ran the cell

Same.

If you ever have this problem again it may be due to a dropped VSCode connection or similar; reloading the window might help. Is it a problem you regularly have?

**Comment by MargaretDuff** [comment]
> > Just a note - this worked fine for me - I was able to see the progress bar both times I ran the cell
> 
> Same.
> 
> If you ever have this problem again it may be due to a dropped VSCode connection or similar; reloading the window might help. Is it a problem you regularly have?

Could we use `run(callback=[TextProgressCallback()])`? Some text like: we use the CIL `TextProgressCallback` to print out information every `update_objective_interval` so you can more clearly see what is happening. For more information on callbacks see https://tomographicimaging.github.io/CIL/nightly/optimisation/#callbacks

**Comment by hrobarts** [comment]
> > > Just a note - this worked fine for me - I was able to see the progress bar both times I ran the cell
> > 
> > 
> > Same.
> > If you ever have this problem again it may be due to a dropped VSCode connection or similar; reloading the window might help. Is it a problem you regularly have?
> 
> Could we use `run(callback=[TextProgressCallback()])`? Some text like: we use the CIL `TextProgressCallback` to print out information every `update_objective_interval` so you can more clearly see what is happening. For more information on callbacks see https://tomographicimaging.github.io/CIL/nightly/optimisation/#callbacks

Thank you for looking at this! Adding the callbacks seems nice, here's the output I get
image

image

I added this text
image

File: demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb (hunk 4 above)
Comment (hrobarts) [reply demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb:372]: I think we can use callbacks to allow early stopping if tolerance is reached

Comment (MargaretDuff) [reply demos/1_Introduction/04_FBP_CGLS_SIRT.ipynb:372]: It will be deprecated in 24.2 I think... and will be replaced by a callback

**Review by lauramurgatroyd — APPROVED** [review]

**Comment by lauramurgatroyd** [comment]
Shall we merge?

**Comment by lauramurgatroyd** [comment]
Or are we waiting to merge all the v24 in one go?

**Comment by hrobarts** [comment]
This now won't work without the callbacks so I think we need to wait until everything is ready then update the required version to 24.1

**Comment by MargaretDuff** [comment]
Let's do it in one go - there are changes in 24.1 that break backwards compatibility so we should do all at once and update the install instructions at the same time

**Comment by MargaretDuff** [comment]
Tested on 24.2.0 and all seems ok
